experimental/Schemes/CoveredScheme.jl

```diff
@@ -0,0 +1,839 @@
+
+### type constructors
+covered_scheme_type(::Type{T}) where {T<:Spec} = CoveredScheme{covering_type(T), morphism_type(covering_type(T))}
+covered_scheme_type(X::Spec) = covered_scheme_type(typeof(X))
+
+covered_scheme_type(::Type{T}) where {T<:ProjectiveScheme} = covered_scheme_type(affine_patch_type(P))
+covered_scheme_type(P::ProjectiveScheme) = covered_scheme_type(typeof(P))
+
+### getter methods
+coverings(X::CoveredScheme) = X.coverings
+refinements(X::CoveredScheme) = X.refinements
+getindex(X::CoveredScheme, C::CoveringType, D::CoveringType) where {CoveringType<:Covering} = X.refinements[(C, D)]
+setindex(X::CoveredScheme, f::CoveringMorphismType, C::CoveringType, D::CoveringType) where {CoveringMorphismType<:CoveringMorphism, CoveringType<:Covering} = X.refinements[(C, D)]
+default_covering(X::CoveredScheme) = X.default_covering
+getindex(X::CoveredScheme, i::Int) = coverings(X)[i]
+
+patches(X::CoveredScheme) = patches(default_covering(X))
+glueings(X::CoveredScheme) = glueings(default_covering(X))
+
+function getindex(X::CoveredScheme, C::Covering)
+  for i in 1:length(coverings(X))
+    C == coverings(X)[i] && return i
+  end
+  error("covering not listed")
+end
+
+getindex(X::CoveredScheme, i::Int, j::Int) = X[X[i], X[j]]
+
+
+set_name!(X::CoveredScheme, name::String) = set_attribute!(X, :name, name)
+name_of(X::CoveredScheme) = get_attribute(X, :name)::String
+has_name(X::CoveredScheme) = has_attribute(X, :name)
+
+function dim(X::CoveredScheme) 
+  if !has_attribute(X, :dim)
+    d = -1
+    is_equidimensional=true
+    for U in patches(default_covering(X))
+      e = dim(U)
+      if e > d
+        d == -1 || (is_equidimensional=false)
+        d = e
+      end
+    end
+    set_attribute!(X, :dim, d)
+    if !is_equidimensional
+      # the above is not an honest check for equidimensionality,
+      # because in each chart the output of `dim` is only the 
+      # supremum of all components. Thus we can only infer 
+      # non-equidimensionality in case this is already visible
+      # from comparing the diffent charts
+      set_attribute(X, :is_equidimensional, false)
+    end
+  end
+  return get_attribute(X, :dim)::Int
+end
+
+function set_default_covering!(X::CoveredScheme, C::Covering) 
+  C in coverings(X) || error("covering is not listed")
+  X.default_covering = C
+  return X
+end
+
+### constructors 
+
+function CoveredScheme(C::Covering)
+  refinements = Dict{Tuple{typeof(C), typeof(C)}, morphism_type(C)}()
+  X = CoveredScheme([C], refinements)
+  set_attribute!(X, :seed_covering, C)
+  return X
+end
+
+CoveredScheme(X::Spec) = CoveredScheme(Covering(X))
+
+# construct the empty covered scheme over the ring R
+function empty_covered_scheme(R::RT) where {RT<:AbstractAlgebra.Ring}
+  return CoveredScheme(empty_spec(R))
+end
+
+
+function Base.show(io::IO, X::CoveredScheme)
+  if has_name(X)
+    print(io, name_of(X))
+    return
+  end
```

Comment:
With the naming system in AbstractAlgebra, one would instead do this:
```suggestion
  @show_name(io, X)
```
However, I am not sure it should be used here... It has some funky code to "guess" names and personally I am ambivalent about it...